docs/architecture/adr-015-ibc-packet-receiver.md

```diff
@@ -0,0 +1,292 @@
+# ADR 015: IBC Packet Receiver
+
+## Changelog
+
+- 2019 Oct 22: Initial Draft
+
+## Context
+ 
+[ICS 26 - Routing Module](https://github.com/cosmos/ics/tree/master/spec/ics-026-routing-module) defines a function [`handlePacketRecv`](https://github.com/cosmos/ics/tree/master/spec/ics-026-routing-module#packet-relay).
+
+In ICS 26, the routing module is defined as a layer above each application module
+which verifies and routes messages to the destination modules. It is possible to 
+implement it as a separate module, however, we already have functionality to route
+messages upon the destination identifiers in the baseapp. This ADR suggests 
+to utilize existing `baseapp.router` to route packets to application modules.
+
+Generally, routing module callbacks have two separate steps in them, 
+verificaton and execution. This corresponds to the `AnteHandler`-`Handler`
+model inside the SDK. We can do the verification inside the `AnteHandler`
+in order to increase developer ergonomics by reducing boilerplate 
+verification code.
+
+## Decision
+
+`PortKeeper` will have the capability key that is able to access only the 
+channels bound to the port. Entities that hold a `PortKeeper` will be
+able to call the corresponding `ChannelKeeper`s method, but only with the
+allowed port. `ChannelKeeper.Port(string, ChannelChecker)` will be defined to
+easily construct a capability-safe `PortKeeper`. This will be addressed in 
+another ADR and we will use unsecure `ChannelKeeper` for now.
+
+`baseapp.runMsgs` will break the loop over the messages if one of the handlers
+returns `!Result.IsOK()`. However, the outer logic will write the cached 
+store if `Result.IsOK() || Result.Code.IsBreak()`. `Result.Code.IsBreak()` if
+`Result.Code == CodeTxBreak`.
+
+```go
+// Pseudocode
+func (app *BaseApp) runTx(tx sdk.Tx) (result sdk.Result) {
+  msgs := tx.GetMsgs()
+  
+  // AnteHandler
+  if app.anteHandler != nil {
+    anteCtx, msCache := app.cacheTxContext(ctx)
+    newCtx, err := app.anteHandler(anteCtx, tx)
+    if !newCtx.IsZero() {
+      ctx = newCtx.WithMultiStore(ms)
+    }
+    
+    if err != nil {
+      // error handling logic
+      return res
+    }
+    
+    msCache.Write()
```

Comment:
where is this defined?

Reply:
It is defined at https://github.com/cosmos/cosmos-sdk/blob/master/baseapp/baseapp.go#L611